### PR TITLE
CST-160: Notify SITs of participant withdrawal

### PR DESCRIPTION
**Pull request by xswoodsx**
## Ticket and context

Ticket: [160](https://dfedigital.atlassian.net/jira/software/projects/CST/boards/119?assignee=5f3543bb3e9e2e004db3189d&selectedIssue=CST-160)

We need to notify SITs when their provider has withdrawn a participant. 

### Code quality checks
- [ ] All commit messages are meaningful and true
- [ ] Added enough automated tests

### HTML Checks
- [ ] All new pages have automated accessibility checks
- [ ] All new pages have visual tests via Percy

### Gotchas
- [ ] All `School` queries are correctly scoped - `eligible` when they need to be

### CODE REVIEW

File: app/services/participants/withdraw/send_withdrawn_email.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        @participant = participant
+        @induction_coordinator = induction_coordinator
+      end
+
+      def call
+        SchoolMailer.fip_provider_has_withdrawn_a_participant(withdrawn_participant: participant, induction_coordinator: induction_coordinator).deliver_later
+      end
+    end
+  end
+end
```
Comment (BroiSatse) [inline app/services/participants/withdraw/send_withdrawn_email.rb:18]: I am not convinced we need a separate service to call one method on the mailer

File: app/mailers/school_mailer.rb [hunk 2]
```diff
@@ -408,4 +409,22 @@ def sit_fip_participant_validation_deadline_reminder_email(induction_coordinator
...
+      rails_mailer: mailer_name,
+      rails_mail_template: action_name,
+      personalisation: {
+        name: induction_coordinator.user.full_name,
+        withdrawn_participant_name: withdrawn_participant.user.full_name,
+        school: withdrawn_participant.school.name,
+        delivery_partner: partnership.delivery_partner.name,
+        lead_provider: partnership.lead_provider.name,
+      },
+    ).tag(:sit_fip_provider_has_withdrawn_a_participant).associate_with(induction_coordinator, as: :induction_coordinator)
```
Comment (BroiSatse) [inline app/mailers/school_mailer.rb:428]: WE'd also would like to associate it with withdrawn_participant

File: spec/services/participants/withdraw/send_withdrawn_email_spec.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  describe "#call" do
+    it "calls SchoolMailer.fip_provider_has_withdrawn_a_participant" do
+      allow(SchoolMailer).to receive(:fip_provider_has_withdrawn_a_participant).and_return(mailer)
+
+      described_class.new(participant: participant_profile, induction_coordinator: sit_profile).call
+
+      expect(SchoolMailer).to have_received(:fip_provider_has_withdrawn_a_participant).with(
+        withdrawn_participant: participant_profile,
+        induction_coordinator: sit_profile,
+      )
```
Comment (BroiSatse) [inline spec/services/participants/withdraw/send_withdrawn_email_spec.rb:21]: Would be great to test that `deliver_later` was called as well.

File: spec/shared/context/lead_provider_profiles_and_courses.rb [hunk 4]
```diff
@@ -11,7 +11,7 @@
   let(:ecf_lead_provider) { cpd_lead_provider.lead_provider }
   let!(:ect_profile) { create(:ect_participant_profile, schedule: default_schedule) }
   let!(:mentor_profile) { create(:mentor_participant_profile, school_cohort: ect_profile.school_cohort, schedule: default_schedule) }
-  let(:induction_coordinator_profile) { create(:induction_coordinator_profile) }
+  let(:induction_coordinator_profile) { create(:induction_coordinator_profile, schools: [ect_profile.school]) }
```
Comment (BroiSatse) [inline spec/shared/context/lead_provider_profiles_and_courses.rb:14]: Why this change?

File: spec/support/shared_examples/participant_actions_withdraw_support.rb [hunk 5]
```diff
@@ -9,9 +9,19 @@
   end
 
   it "creates a withdrawn state and makes the profile withdrawn" do
+    mailer = double(SchoolMailer, deliver_later: nil)
```
Comment (BroiSatse) [inline spec/support/shared_examples/participant_actions_withdraw_support.rb:12]: `instance_double`

File: spec/support/shared_examples/participant_actions_withdraw_support.rb [hunk 6]
```diff
@@ -9,9 +9,19 @@
...
+    allow(SchoolMailer).to receive(:fip_provider_has_withdrawn_a_participant).and_return(mailer)
+
     expect { described_class.call(params: given_params) }.to change { ParticipantProfileState.count }.by(1)
     expect(user_profile.participant_profile_state).to be_withdrawn
     expect(user_profile).to be_training_status_withdrawn
+
+    if user_profile.user.participant?
```
Comment (BroiSatse) [inline spec/support/shared_examples/participant_actions_withdraw_support.rb:19]: `user_profile.user`? Should we rename `user_profile` to `participant_profile`?

Comment (xswoodsx) [reply spec/support/shared_examples/participant_actions_withdraw_support.rb:19]: Would that not need a lot of changes? I thought `user_profile` came from https://github.com/DFE-Digital/early-careers-framework/blob/develop/spec/services/participants/withdraw/early_career_teacher_spec.rb

File: spec/shared/context/lead_provider_profiles_and_courses.rb (hunk 4 above)
Comment (xswoodsx) [reply spec/shared/context/lead_provider_profiles_and_courses.rb]: yeah definitely not needed, i left this in.

File: spec/support/shared_examples/participant_actions_withdraw_support.rb (hunk 6 above)
Comment (tonyheadford) [reply spec/support/shared_examples/participant_actions_withdraw_support.rb]: Looks like these declaration specs are riddled with `user_profile` being used as an abstraction instead of `participant_profile`, there may be a reason for it. I'd be tempted to follow the same pattern for now.

**Review by tonyheadford — APPROVED** [review]
LGTM although check that @BroiSatse is OK with the `user_profile` as he raised it.

File: app/mailers/school_mailer.rb [hunk 7]
```diff
@@ -429,4 +430,26 @@ def school_preterm_reminder(school:, season:)
...
+        school: withdrawn_participant.school.name,
+        delivery_partner: partnership.delivery_partner.name,
+        lead_provider: partnership.lead_provider.name,
+      },
+    )
+    email
+      .tag(:sit_fip_provider_has_withdrawn_a_participant)
+      .associate_with(induction_coordinator, as: :induction_coordinator)
+      .associate_with(withdrawn_participant, as: :participant_profile)
+  end
```
Comment (BroiSatse) [inline app/mailers/school_mailer.rb:454]: Note: we really need to stop putting all the emails on the school_mailer... But that a bit of a refactor so let's skip it for now though...

File: app/services/participants/withdraw/validate_and_change_state.rb [hunk 8]
```diff
@@ -16,6 +16,11 @@ def perform_action!
         ActiveRecord::Base.transaction do
           ParticipantProfileState.create!(participant_profile: user_profile, state: ParticipantProfileState.states[:withdrawn], reason: reason)
           user_profile.training_status_withdrawn!
+
+          unless user_profile.npq?
+            induction_coordinator = user_profile.school&.induction_coordinator_profiles&.first
```
Comment (BroiSatse) [inline app/services/participants/withdraw/validate_and_change_state.rb:21]: school is required for non-npq profiles, we should be fine with dropping both `&` here.

File: app/services/participants/withdraw/validate_and_change_state.rb [hunk 9]
```diff
@@ -16,6 +16,11 @@ def perform_action!
...
+            SchoolMailer.fip_provider_has_withdrawn_a_participant(withdrawn_participant: user_profile, induction_coordinator: induction_coordinator).deliver_later
```
Comment (BroiSatse) [inline app/services/participants/withdraw/validate_and_change_state.rb:22]: Since we moved to sidekiq we should always schedule email outside of DB transactions as the worker can pick up scheduled job before the transaction is committed and sending email without all the data,

**Review by BroiSatse — APPROVED** [review]

File: app/services/participants/withdraw/validate_and_change_state.rb (hunk 9 above)
Comment (xswoodsx) [reply app/services/participants/withdraw/validate_and_change_state.rb:22]: Good spot! i noticed Tony did that for Partnership Challenge email on Friday, hadn't spotted it in my own.
